File: amqp_setup.ts
// rabbitmq-mgmt container id: 0c476c4886f8
// localhost:15672 -- rabbitmq

import amqp from "amqplib";
// const open = amqp.connect("amqp://localhost:5672");
// const conn = amqp.connect({
//   protocol: 'amqp',
//   hostname: 'localhost',
//   port: 5672,
//   locale: 'en_US',
//   heartbeat: 3600
// });
const exchange = { name: 'message_topic', type: 'topic' };
const options = { durable: true };
const queue = { name: 'Twilio-Sms', key: '*.notification' };

// conn.then((connection) => connection.createChannel())
//   .then((channel) => {
//     channel.assertExchange(exchange.name, exchange.type, options);
//     channel.assertQueue(queue.name, options);
//     channel.bindQueue(queue.name, exchange.name, queue.key);
//   })
//   .catch(err => { throw err });

async function openChannel() {
  const connection = await amqp.connect({
    protocol: 'amqp',
    hostname: 'localhost',
    port: 5672,
    locale: 'en_US',
    heartbeat: 3600
  });
  const channel = await connection.createChannel();
  return { channel, connection };
};

function closeCh(ch: amqp.Channel, conn: amqp.Connection) {
  setTimeout(() => {
    ch.close();
    conn.close();
  }, 500);
};

// server.ts
const send = async () => {
  const { channel, connection } = await openChannel();
  await channel.assertExchange(exchange.name, exchange.type, options);
  await channel.assertQueue(queue.name, options);
  await channel.bindQueue(queue.name, exchange.name, queue.key);
  await channel.publish(exchange.name, queue.key, Buffer.from(""));
  closeCh(channel, connection);
};

// index.ts
const receive = async () => {};

export { exchange, options, queue, send, receive };

File: amqp_setup.py
# python example is213 amqp for sending and receiving

import pika
from os import environ
import atexit


# set up exchanges
log_exchange = "log_topic"
notif_exchange = "notif_topic"
# setup queues
queues = {
  "error": { "name": "Error_Log", "key": "*.error", "exchange": log_exchange },
  "activity": { "name": "Activity_Log", "key": "*.activity", "exchange": log_exchange },
  "notif": { "name": "Twilio-Sms", "key": "#", "exchange": notif_exchange }
}

rabbitHost = environ.get("rabbit_host") or "localhost"
rabbitPort = environ.get("rabbit_port") or 5672

# open connection and setup queues
connection = pika.BlockingConnection(
  pika.ConnectionParameters(
    host=rabbitHost,
    port=rabbitPort,
    locale="en-US",
    heartbeat=3600
  )
)
channel = connection.channel()
# declare the log and notif exchanges
channel.exchange_declare(exchange=log_exchange, exchange_type="topic", durable=True)
channel.exchange_declare(exchange=notif_exchange, exchange_type="topic", durable=True)


# declare and bind the 3 queues, 2 to log exchanage, 1 to notif exchange
for queue in queues.values():
  channel.queue_declare(queue=queue.get("name"), durable=True)
  channel.queue_bind(exchange=queue.get("exchange"), queue=queue.get("name"), routing_key=queue.get("key"))


# export: helper functions for publishing data to amqp
def send(exchange, queue, data):
  channel.basic_publish(
    exchange=exchange,
    routing_key=queue.get("key"),
    body=data, # data: string, transformed into byte code
    properties=pika.BasicProperties(delivery_mode = 2) # persistent
  )
# helper function for sending an error log
def sendErrorLog(error_log):
  send(log_exchange, queues.get("error"), error_log)
# helper function for sending an activity log
def sendActivityLog(activity_log):
  send(log_exchange, queues.get("activity"), activity_log)
# helper function for sending an sms through Twilio
def sendMsg(smsData):
  send(notif_exchange, queues.get("notif"), smsData)


# closing channel when process is ended
# def closeChannel():
#   channel.close()
#   connection.close()
# atexit.register(closeChannel())